include/wil/safearrays.h

```diff
@@ -0,0 +1,1105 @@
+    //!         bstr = (*current++).release();
+    //!     }
+    //!     return sa;
+    //! }
+    //! ~~~~
+    //! void handle_safearray_of_bools(SAFEARRAY* psa, std::function<void(bool, ULONG)> fnHandler)
+    //! {
+    //!     auto data = wil::safearraydata<VARIANT_BOOL>{psa};
+    //!     for (auto i = 0U; i < data.size(); ++i)
+    //!     {
+    //!         // Translate to Regular Bool and Invoke the Handler
+    //!         fnHandler(data[i] != VARIANT_FALSE, i);
+    //!     }
+    //! }
+    //! ~~~~
+    //! template<typename T>
+    //! std::vector<T> extract_all_values(SAFEARRAY* psa)
+    //! {
+    //!     auto data = wil::safearraydata<T>{psa};
+    //!     // Construct the vector from the iterators
+    //!     return {data.begin(), data.end()};
+    //! }
+    //! ~~~~
+    //! template<typename INTERFACE>
+    //! std::vector<wil::com_ptr_t<INTERFACE, wil::err_exception_policy>> extract_all_interfaces(SAFEARRAY* psa)
+    //! {
+    //!     auto result = std::vector<wil::com_ptr_t<INTERFACE, wil::err_exception_policy>>{};
+    //!     auto data = wil::safearraydata<LPUNKNOWN>{psa};
+    //!     result.reserve(data.size());
+    //!     for (auto& p : data) // Type of p is LPUNKNOWN
+    //!     {
+    //!         // Use "tag_com_query" if you want failure instead of nullptr if INTERFACE not supported
+    //!         result.emplace_back(p, wil::details::tag_try_com_query);
+    //!     }
+    //!     return result;
+    //! }
+    //! ~~~~
+    //! @tparam T           The underlying datatype expected in the SAFEARRAY. It should be a scalar type, BSTR, VARIANT_BOOL, LPUNKNOWN/LPDISPATCH, etc.
+    //! @tparam err_policy  Represents the error policy for the class (error codes, exceptions, or fail fast; see @ref page_errors)
+    template <typename T, typename err_policy = err_exception_policy>
+    class safearraydata_t
+    {
+    public:
+        typedef typename err_policy::result result;
+        typedef safearray_unaccess_scope_exit::pointer pointer;
+        typedef T value_type;
+        typedef value_type* value_pointer;
+        typedef const value_type* const_value_pointer;
+        typedef value_type& value_reference;
+        typedef const value_type& const_value_reference;
+        typedef value_type* iterator;
+        typedef const value_type* const_iterator;
+
+    public:
+        // Exception-based construction
+        explicit safearraydata_t(pointer psa)
+        {
+            static_assert(wistd::is_same<void, result>::value, "this constructor requires exceptions; use the create method");
+            access(psa);
+        }
+        safearraydata_t() = default;
+        safearraydata_t(safearraydata_t&&) = default;
+        ~safearraydata_t() = default;
+        safearraydata_t& operator=(safearraydata_t&&) = default;
+        safearraydata_t(const safearraydata_t&) = delete;
+        safearraydata_t& operator=(const safearraydata_t&) = delete;

```

Comment:
```suggestion
        safearraydata_t() = default;
```
Just to be clearer with my suggestion above